### ISSUE TITLE
[Feature Request] Add 'Generate password using current pattern' option

### ISSUE THREAD

**Issue by ba32107** [keepassxreboot/keepassxc#2628]
When generating a new password for an existing entry, I would like an option that generates a new password using the existing settings for the given entry.

Some sites only allow alphanumerical characters and/or has a max character limit. This feature would allow to easiily generate a new password for any site using different patterns.

Thanks,
Balazs

**Comment by droidmonkey** [comment]
This was discussed before. The problem with this is you have to store the current profile for each entry in order to replicate it exactly each time. This is better handled through pre-defined and customizable password profiles that are available database-wide. We cannot "guess" what password settings you are using based on your current password since it is random.

**Comment by joshualibrarian** [comment]
Even if "pre-defined password profiles are available database-wide", that doesn't really solve the problem presented in (closed) #3091 (https://github.com/keepassxreboot/keepassxc/issues/3091).  Whatever profile I started with, when I go to generate a password for a new site, I invariably have to customize the generator settings.  Just now, I had to select the right boxes and then paste the list of "additional allowed characters" into the allowed box.  These generator settings should be preserved and associated with THAT entry so that next time I go to change the password for that entry, the generator is already set correctly.  Is it not viable to store this data in the database with the entry, whether as reference to database-wide profile or not?

**Comment by droidmonkey** [comment]
Also allow setting a default configuration that is easily chosen.

**Comment by GenocideStomper** [comment]
I would also love to see password profiles for the "Password Generator" tool. Sometimes I need to change settings, and don't want to lose the characters from the "Also choose from:" and "Do not include:" fields.

**Comment by Massimo-B** [comment]
![Screenshot_2022-10-13-08-42-28-320_keepass2android keepass2android-crop](https://user-images.githubusercontent.com/3286728/195583019-1c61e0a3-f252-418f-a4d4-e72214ae0251.jpg)
Just to give an idea here is a screenshot of the keepass2android application Password Generator with profiles.

**Comment by mattesony** [comment]
I really like this feature and miss having it in my previous password manager. I'd be happy to take a stab at it either with the profiles or by storing the generating character set/length on an entry-by-entry basis. I think it could be built on by adding some form of inheritance of the character set/length from the groups the entries are in if the character set/length is unset for the entry, then the profiles would just be a further refinement to allow easier management.

**Comment by michaelk83** [comment]
The profiles should be compatible with OG KeePass. As I recall, there was a specification of it somewhere.

**Comment by droidmonkey** [comment]
This should being implemented exactly how saved searches was implemented, except of course in the password generator. I don't think this is in the kdbx specifications.

**Comment by michaelk83** [comment]
I think I saw some sort of spec or similar document on the OG KeePass website a while back, but I can't find it now. In any case, it does have [password generator profiles](https://keepass.info/help/base/pwgenerator.html#profiles), and so does Keepass2Android, as [noted above](https://github.com/keepassxreboot/keepassxc/issues/2628#issuecomment-1277448960). So KeePassXC shouldn't reinvent its own system for this.

**Comment by droidmonkey** [comment]
Depends on how much I like the way they implemented it. Our password generator is decidedly different then KeePass2 and KeePass2Android actually uses the KeePass2 binaries under the hood.

**Comment by michaelk83** [comment]
It could be an issue for some users if they make password profiles in OG KeePass (or on android), and can't use them in KeePassXC, or vice versa. Though currently that is already the case (since KeePassXC doesn't support profiles at all).

If your new way is that much better, I suppose it can be standardized in KDBX5.

**Comment by Thunderhawk2001** [comment]
Any news on this?

And i hope this is the right place for this: 
It would be awesome, if it would be possible to add random numbers to the passphrase: insert a number (0-9) randomly in front or back of a word and/or after a separator. e.g. "word-word6-word" or "word-6word-word"

**Comment by Massimo-B** [comment]
On Jan 20, 2019, [droidmonkey](https://github.com/droidmonkey) changed the title from '_Generate password using current pattern_' to '_Add ability to customize password profiles for each database_'. I think this does not match the initial feature request anymore, which was "_generates a new password using the existing settings for the given entry_". Initially this meant to have a password profile stored with the entry not with the database.

The issue that needs to be solved, according to my opinion, I can't remember every single websites password policy to re-generate a new password. On most websites it doesn't matter but if I got some buggy website and finally figured out a working password profile, I would like to store that with the entry.

**Comment by Massimo-B** [comment]
> And i hope this is the right place for this: It would be awesome, if it would be possible to add random numbers to the passphrase: insert a number (0-9) randomly in front or back of a word and/or after a separator. e.g. "word-word6-word" or "word-6word-word"

I think this is a completely different request. You want to have a pattern like a Regular Expression or something that has fixed strings, variable character sets and counts and stuff. I don't think any website has policies requiring passwords based on such patterns. It might help you remembering passwords, but right after starting with password managers, strong unique passkeys are the only right way to go... my opinion.

**Comment by Massimo-B** [comment]
> Even if pre-defined password profiles are available database-wide, that doesn't really solve the problem presented in (closed) #3091.

I agree. Currently KeePassXC does not have any kind of persistent password profiles. But to solve this feature request such a profile would need to be stored together with the password entry itself and obviously not guessed from the random password.

**Comment by Thunderhawk2001** [comment]
> > And i hope this is the right place for this: It would be awesome, if it would be possible to add random numbers to the passphrase: insert a number (0-9) randomly in front or back of a word and/or after a separator. e.g. "word-word6-word" or "word-6word-word"
> 
> I think this is a completely different request. You want to have a pattern like a Regular Expression or something that has fixed strings, variable character sets and counts and stuff. I don't think any website has policies requiring passwords based on such patterns. It might help you remembering passwords, but right after starting with password managers, strong unique passkeys are the only right way to go... my opinion.

I mean something like Bitwarden Password Generator does it:
"Just" a toggle for "include number"
![grafik](https://github.com/user-attachments/assets/17d594ba-2a81-45c7-a2fc-0cd3e7ce7610)


Sorry, really should drink my coffee before posting stuff.

**Comment by Massimo-B** [comment]
[offtopic] The Passphrase password generation is possible already:
![image](https://github.com/user-attachments/assets/8f374ba4-a793-4a06-9aa9-6b22604b3b33)
Just the random number part not... but anyway, different request.

**Comment by fireba11** [comment]
Providing the option to save/load generator profiles (possibly independent of database) like keepass offers would be great.
That way you could roll out profiles according to regulations to your user.
"Admin password", "API password" etc.

**Comment by Massimo-B** [comment]
And additionally adding a special attribute to an entry to select such a profile would help every time the password needs to be renewed.

**Comment by Grocel** [comment]
This is something I really miss in KeepassXC. In OG Keepass the pattern settings are stored in the global config file, which is actually quite convenient, but also not that secure. KeepassXC could do this much better by using a custom database entry to store the password patterns.

To be honest, I don't get the where the problem with this feature lies beside acting the classic "Developer's No™".

**Comment by droidmonkey** [comment]
There is no problem, it just needs to be developed

**Comment by hgulfbmito** [comment]
From a user perspective, a common profile for all accounts would actually lead to less security, as users would simply use the smallest subset. If it's not part of the DB format, it could either become one. A second password profile file with the same name would be another option. Coordination with Dominik Reichl (the developer of the original program) would probably be a good idea to maintain compatibility with any other programs that use the same DB format.
People need a PW manager because they have to deal with all the "stupid password rules which require character X to be present and disallow character Y", and, unfortunately, these PW rules are the norm rather than the exception.

**Comment by thethinker990** [comment]
+1 for password profiles like private, work, API, admin and so on (independent of any database or entries)

**Comment by Massimo-B** [comment]
I rather need it per entry, not per database. Because when switching passwords I don't remember if sites have policies or bugs with 30+ characters or special groups. So once found the specific policy I would save it to the single entry.

**Comment by droidmonkey** [comment]
The database level holds custom password profiles, the entry level stores which profile was used last (if any).

**Comment by CaptainPalapa** [comment]
Asking Claude about the ability to save password generation profiles, it led me to Issue #9124 and that was marked as a dupe of this.  I, too, would very much like to see that feature added. I VERY commonly use these in KeePass but I'm dropping it in favor of the XC (and DX on the phone).

- 20: Upper+Lower+Number+$@, length 20
- 32: Upper+Lower+Number+$@, length 32
- 50: Upper+Lower+Number+$@, length 50

If I can use 50, I do, 32 more often.

**Comment by Massimo-B** [comment]
The initial idea was to store the profile with the entry. Because it's often hard to find out which password policy is accepted by sites, and only to be found out by long tests with different sizes and character groups. It would be useful to save this profile with the entry, for later password renewals of a long list of entries it would be highly useful to not do all the testing again. For now I need to store the policy as a comment in the entry.

**Comment by Grocel** [comment]
Why not both?

**Comment by droidmonkey** [comment]
When this is implemented, the password profiles will be stored in the database custom data and entries MAY reference the profile that was used (in the entry custom data) to generate the current password.

**Comment by burtek** [comment]
Since some sites ban certain characters, will the profiles also keep the allowlist and banlist (a.k.a 'Also choose from' and 'Do not include')?

**Comment by arnosec** [comment]
@droidmonkey  with this you closed my issue, but also add a direct regenerate password button next to dice button in the password field in case we just want to change password directly without modifying the dice menu options.

**Comment by pederbe** [comment]
@droidmonkey  I'd be interested in working on this. 

I found your existing backend PR, https://github.com/droidmonkey/keepassxc/pull/11, and saw your clarification about storing profiles in database custom data, with entries optionally referencing the profile used.

Is there any newer work elsewhere that I should know about? If not, would you be open to me building on that backend and tackling the generator UI and entry integration?

**Comment by droidmonkey** [comment]
You can use the prior work as a strong start.
